graphql_compiler/macros/macro_expansion.py

```diff
@@ -28,21 +29,144 @@ def _merge_non_overlapping_dicts(merge_target, new_data):
+    selection_dict_a = get_uniquely_named_objects_by_name(selection_set_a.selections)
+    selection_dict_b = get_uniquely_named_objects_by_name(selection_set_b.selections)
+
+    # Compute intersection by name
+    common_selection_dict = dict()
+    common_fields = set(selection_dict_a.keys()) & set(selection_dict_b.keys())
+    for field_name in common_fields:
+        field_a = selection_dict_a[field_name]
+        field_b = selection_dict_b[field_name]
+        if field_a.selection_set is not None or field_b.selection_set is not None:
+            raise GraphQLCompilationError('TODO')
+
+        merged_field = deepcopy(field_a)
+        merged_field.directives += field_b.directives
+        common_selection_dict[field_name] = merged_field
+
+    # Merge dicts
+    merged_selection_dict = copy(selection_dict_a)
+    merged_selection_dict.update(selection_dict_b)
+    merged_selection_dict.update(common_selection_dict)
+
+    # Remove pro-forma fields if allowed
+    if len(merged_selection_dict) > 1:
+        merged_selection_dict = {
+            name: ast
+            for name, ast in six.iteritems(merged_selection_dict)
+            if ast.selection_set is not None or len(ast.directives) > 0
+        }
+
+    # Get a deterministic ordering of the merged selections
+    selection_name_order = [
+        ast.name.value
+        for ast in selection_set_a.selections
+        if ast.name.value not in selection_dict_b
+    ] + [
+        ast.name.value
+        for ast in selection_set_b.selections
+    ]
+
+    # Make sure that all property fields come before all vertex fields.
+    return SelectionSet([
+        merged_selection_dict[name]
+        for name in selection_name_order
+        if name in merged_selection_dict and merged_selection_dict[name].selection_set is None
+    ] + [
+        merged_selection_dict[name]
+        for name in selection_name_order
+        if name in merged_selection_dict and merged_selection_dict[name].selection_set is not None
+    ])
+
+
+def _merge_selection_into_target(target_ast, selection_ast, subclass_sets=None):
+    """Add the selections, directives, and coercions from the selection_ast to the target_ast."""
+    # Deal with coercions at the target
+    continuation_ast = selection_ast
+    first_selection = selection_ast.selection_set.selections[0]
+    if isinstance(first_selection, InlineFragment):
+        coercion_class = first_selection.type_condition.name.value
+        if isinstance(target_ast, InlineFragment):
+            target_class = target_ast.type_condition.name.value
+            continuation_ast = first_selection
+            target_ast.type_condition = first_selection.type_condition
+
+            # coercion_class is required to subclass target_class so we can merge the type coercions
+            if coercion_class != target_class:
+                if subclass_sets is None:
+                    raise AssertionError(u'Cannot prove type coercion at macro target is valid. '
```

Comment:
Is this something a user can trigger with an invalid use of a macro, or is it always a configuration error? `AssertionError` is fine for the latter, but not fine if a user can cause it with a bad GraphQL query on a correctly-configured system.